src/main/resources/io/viash/platforms/nextflow/workflowFactory/processFactory.nf

```diff
@@ -0,0 +1,210 @@
+// depends on: thisConfig, thisScript, session?
+def processFactory(Map processArgs) {
+  // autodetect process key
+  def wfKey = processArgs["key"]
+  def procKeyPrefix = "${wfKey}_process"
+  def meta = nextflow.script.ScriptMeta.current()
+  def existing = meta.getProcessNames().findAll{it.startsWith(procKeyPrefix)}
+  def numbers = existing.collect{it.replace(procKeyPrefix, "0").toInteger()}
+  def newNumber = (numbers + [-1]).max() + 1
+
+  def procKey = newNumber == 0 ? procKeyPrefix : "$procKeyPrefix$newNumber"
+
+  if (newNumber > 0) {
+    log.warn "Key for module '${wfKey}' is duplicated.\n",
+      "If you run a component multiple times in the same workflow,\n" +
+      "it's recommended you set a unique key for every call,\n" +
+      "for example: ${wfKey}.run(key: \"foo\")."
+  }
+
+  // subset directives and convert to list of tuples
+  def drctv = processArgs.directives
+
+  // TODO: unit test the two commands below
+  // convert publish array into tags
+  def valueToStr = { val ->
+    // ignore closures
+    if (val instanceof CharSequence) {
+      if (!val.matches('^[{].*[}]$')) {
+        '"' + val + '"'
+      } else {
+        val
+      }
+    } else if (val instanceof List) {
+      "[" + val.collect{valueToStr(it)}.join(", ") + "]"
+    } else if (val instanceof Map) {
+      "[" + val.collect{k, v -> k + ": " + valueToStr(v)}.join(", ") + "]"
+    } else {
+      val.inspect()
+    }
+  }
+
+  // multiple entries allowed: label, publishdir
+  def drctvStrs = drctv.collect { key, value ->
+    if (key in ["label", "publishDir"]) {
+      value.collect{ val ->
+        if (val instanceof Map) {
+          "\n$key " + val.collect{ k, v -> k + ": " + valueToStr(v) }.join(", ")
+        } else if (val == null) {
+          ""
+        } else {
+          "\n$key " + valueToStr(val)
+        }
+      }.join()
+    } else if (value instanceof Map) {
+      "\n$key " + value.collect{ k, v -> k + ": " + valueToStr(v) }.join(", ")
+    } else {
+      "\n$key " + valueToStr(value)
+    }
+  }.join()
+
+  def inputPaths = thisConfig.functionality.allArguments
+    .findAll { it.type == "file" && it.direction == "input" }
+    .collect { ', path(viash_par_' + it.plainName + ')' }
+    .join()
+
+  def outputPaths = thisConfig.functionality.allArguments
+    .findAll { it.type == "file" && it.direction == "output" }
+    .collect { par ->
+      // insert dummy into every output (see nextflow-io/nextflow#2678)
+      if (!par.multiple) {
+        ', path{[".exitcode", args.' + par.plainName + ']}'
+      } else {
+        ', path{[".exitcode"] + args.' + par.plainName + '}'
+      }
+    }
+    .join()
+
+  // TODO: move this functionality somewhere else?
```

Comment:
Since we are moving code around, what about this TODO?

Reply:
Hmmm... I'm going to pass on changing this code atm ^^